### ISSUE TITLE
Qualify MCG Trellis K6 and projection-mixed GLM-5.2 serving

### ISSUE THREAD

**Issue by voipmonitor** [voipmonitor/rtx6kpro#73]
## Purpose

GLM-5.2 checkpoints serialized in EXL3 format can store routed-expert
projections with MCG Trellis K3, K4, and K5 payloads and can encode eligible
BF16 dense projections online as unpaired MCG Trellis K6. MCG Trellis is the
quantization and execution contract; K3 through K6 identify the number of
encoded bits per coefficient.

The required B12X and vLLM changes preserve compressed weights throughout
loading and serving. They do not reconstruct routed-expert weights in BF16 and
do not copy ExLlamaV3 implementation code into B12X.

## Status

| Responsibility | Status | Pull request |
|---|---|---|
| Fused unpaired MCG K6 dense projection on SM120 | **implemented and qualified** | [B12X #221](https://github.com/local-inference-lab/b12x/pull/221), head `413f96e889dad1ae0752fd1f4be9d37f56849600` |
| Projection-mixed MCG K3/K4/K5 routed experts and native BTX ownership | **implemented and qualified** | [B12X #223](https://github.com/local-inference-lab/b12x/pull/223), review head `3df80ee3` |
| Projection-mixed EXL3 loading, online K6 cache, and graph storage | **implemented and qualified** | [vLLM #300](https://github.com/local-inference-lab/vllm/pull/300), head `1b0c1c49f3517406db33f9b08aa7b476ce8d3e34` |
| TP4/DCP1/MTP3 serving from an immutable image | **qualified** | Infernal Invocation r17 |

All three pull requests remain open. No source pull request was merged as part
of the r17 release composition.

## Merge contract

1. Merge B12X #221 and B12X #223 into `b12x/master`. The pull requests have
   separate responsibilities and can be reviewed independently.
2. Refresh vLLM #300 against `dev/infernal-invocation` after both B12X APIs are
   available from `b12x/master`.
3. Merge vLLM #300 only after its dependency import and focused source tests
   pass against the merged B12X revision.

B12X #221 provides the fused dense K6 kernel and exact graph-scratch query.
B12X #223 provides projection-mixed routed-expert execution and
`fused_moe.adopt_btx_weights(plan, prepared)`. vLLM #300 consumes both APIs.

### Dense projections

Eligible BF16 dense and shared-expert projections use this operation:

```text
fp16(x * input_sign) -> normalized H128 -> K6/MCG GEMM
-> normalized H128 -> fp16(* output_sign)
```

B12X #221 executes both transforms and the compressed GEMM in one cooperative
CuTe DSL grid. Payload validation excludes paired QSRT records and unsupported
geometry before compilation.

### Routed experts

B12X #223 keeps gate, up, and down projection tier maps independent, preserves
global expert identifiers, and executes K3/K4 and K3/K4/K5 layouts through one
cooperative W4A16 grid. Prepared native BTX storage crosses the vLLM adapter by
ownership transfer instead of tensor copying or repacking.

Route-pack warmup state is keyed by the complete launch contract: device,
route-ID dtype, token count, top-k width, mapping mode, and caller-owned arena
capacities. Different launch shapes cannot reuse an incompatible warmup entry.

### Checkpoint loading

vLLM #300 activates the projection-mixed loader only for checkpoints declaring
`r7_routed_experts`. It validates all packed extents and per-projection tiers,
binds immutable B12X artifacts once per layer, owns CUDA graph scratch, and
prewarms scheduler-reachable target and MTP route shapes.

Eligible BF16 projections use a content-addressed per-rank online K6 cache.
The canonical helper setting is:

```text
ONLINE_QUANT=trellis-mcg-b6
```

`ONLINE_QUANT=exl3-b6` remains a compatibility alias. EXL3 identifies the
checkpoint serialization format and is not the quantization algorithm name.

## Source qualification

The focused B12X suites passed on an NVIDIA RTX PRO 6000 Blackwell Server
Edition GPU (SM120). Coverage includes:

- K3/K4 and K3/K4/K5 numerical parity;
- eager execution and CUDA graph replay;
- projection-tight storage and interleaved global expert routes;
- zero-copy native BTX adoption and serving equivalence;
- route-capacity prewarm and independent cache-key dimensions;
- K6 payload admission, scratch sizing, numerical parity, and graph replay.

The focused vLLM suites passed. Coverage includes:

- checkpoint schema and projection-tier validation;
- uniform and projection-mixed native BTX preparation;
- online K6 cache identity and atomic publication;
- graph-owned scratch and target/MTP warmup;
- disabled-MTP weight handling and retained-storage lifetime.

Ruff check and format, Python compilation, and `git diff --check` passed for
both source trees.

B12X #223 commit `3df80ee3` adds seven cache-identity tests after the r17
runtime qualification. It changes no runtime source. The image contains the
qualified runtime code at `35c7ed0b7ac2430ceb8d07a6fb05734ac5f3878e`.

## Serving qualification

The qualified artifact is:

```text
image:           voipmonitor/vllm:infernal-invocation-vllmc53cc73-b12xc0a44a1-fi1ac6942-cu133-torch213-20260817-r17
registry digest: sha256:c5e96c5bcc5a073f7ce6b56173d88538de3a416900cff97c88b4bf7967fe1dc0
vLLM base:       d6e0bb7c813264f1293288b9de0d18f277be7c9f
vLLM tree:       c53cc73dd64992f013842edf53513f604457c402
B12X base:       c25cdba2c1df7a69b2d7771e4243e12a8fbf19d5
B12X tree:       c0a44a16f884222faab1ff52ed5db0875bf61971
checkpoint:      brandonmusic/GLM-5.2-EXL3-TR3v4-3.5bpw-MTP78@9ab9579774cc432df91567a36f6e9e863e0d4c9f
hardware:        4 x RTX PRO 6000 Blackwell, direct-attached PCIe
runtime:         TP4/DCP1/MTP3, online MCG K6, NVFP4 DS-MLA KV
scheduler:       MNS1, graph cap 4, maximum model length 32,768
```

The container used only packaged source. The model loaded 86.61 GiB per rank,
reused 444 online K6 cache entries, observed K3/K4/K5 projection tiers,
captured FULL and PIECEWISE prefill graphs and a FULL decode graph, and returned
a coherent chat completion. No traceback, CUDA runtime error, GPU Xid event,
or engine-fatal event was observed.

| CC1 metric | Result |
|---|---:|
| Aggregate decode | 94.06 tok/s |
| Active-user decode | 94.35 tok/s |
| Target steps | 33.42 steps/s |
| Effective acceptance length | 2.814 |
| Speculative acceptance | 60.48% |
| Request errors | 0 |

Infernal Invocation r16 measured 33.33 target steps/s with the same checkpoint,
host, topology, and serving configuration. The 0.27% target-step difference is
within run variation. The emitted-token difference follows speculative
acceptance variation and is not evidence of a kernel speedup.

The release source locks, compose recipe, and raw benchmark are stored in
[`blackwell-llm-docker`](https://github.com/local-inference-lab/blackwell-llm-docker/tree/main/patches/releases/infernal-invocation-r17)
and
[`validation/infernal-invocation-r17`](https://github.com/local-inference-lab/blackwell-llm-docker/tree/main/validation/infernal-invocation-r17).

## Qualification boundary

The serving result qualifies TP4/DCP1/MTP3 with one active sequence and a
32,768-token model-length envelope. Larger scheduler capacities, longer
contexts, DCP greater than one, TP8 GLM-5.2 NVFP4, and other checkpoints remain
outside this serving qualification until measured independently.

**Comment by brandonmmusic-max** [comment]
## TP4/DCP4 R7 long-context qualification update

I now have a sealed qualification package for the 3.5 bpw GLM-5.2 checkpoint
on the dedicated, projection-exact R7 graph path. This extends the earlier
TP4/DCP1/32K boundary; it does not replace or broaden other r17 qualification
claims.

### Scope of the requested candidate validation

The requested candidate qualification was performed with **TP4/DCP4/MTP3,
`nvfp4_ds_mla` KV, calibrated static latent scales, BF16 RoPE, and full CUDA
graphs**. Unless a section explicitly says otherwise, every serving result
below—KV capacity, prefill, decode, LAVD, Estonia, Hotel, Estonia-long, and
260K retrieval—belongs to that NVFP4/DCP4 candidate.

The sealed `0.0597302107` KLD result is also included because it is the strongest
numerical quality result, but it is a **separate FP8 MLA reader panel**. It must
not be relabeled as the NVFP4/DCP4 candidate's KLD. The NVFP4 numerical record
is reported separately below: one lower completed observation at
`0.0639764262` and a four-observation descriptive mean of `0.0691655450`.

### Separate headline quality result: sealed FP8 MLA KLD gate

The most important numerical result is the completed exact-G64-Q FP8 MLA
reader panel. All five fresh-boot observations completed against the same
official BF16-reference logits:

| Run | Mean KLD |
|---:|---:|
| 1 | 0.0585513921 |
| 2 | 0.0638344169 |
| 3 | 0.0589407659 |
| 4 | 0.0592789936 |
| 5 | 0.0580454851 |
| **Five-run arithmetic mean** | **0.0597302107** |

**Verdict: PASS.** The predeclared acceptance gate was a five-fresh-boot
arithmetic mean below `0.060`; the measured mean was `0.0597302107`. The panel
used FP8 MLA KV with BF16 RoPE and exact FP32 Q scaling at group width 64 on
TP4/DCP4. This is a completed five-run result, not an estimate derived from a
partial panel.

The FP8 panel used the Infernal Invocation r17 image
`voipmonitor/vllm:infernal-invocation-vllmc53cc73-b12xc0a44a1-fi1ac6942-cu133-torch213-20260817-r17`
at base digest
`sha256:c5e96c5bcc5a073f7ce6b56173d88538de3a416900cff97c88b4bf7967fe1dc0`.
The token-ID SHA-256 was
`d0267e72233a0e367768a09d989dd4bd6ba5f18dfb485e198b71cface1333056`;
the BF16-reference-logit SHA-256 was
`87f992a689c054a0548a4b3863da6c809f9239beacd5786d0401e45904fec063`.
All five full-logit receipts and the machine summary are included in the
review package.

An optional comparison with the five-run stock-control mean (`0.0603338410`)
does not resolve a population-level difference with only five boots. That is a
separate statistical question and does not alter the declared gate result.

### NVFP4 graph artifact and runtime

- Model: `brandonmusic/GLM-5.2-EXL3-TR3v4-3.5bpw-MTP78`
- Task/performance image: `sha256:c4dfd179d5cf40fe4643c15bbd7164fee1a3d8d39d2d4d514867c74bfe7624f5`
- KLD receipt image: `sha256:6de02ac557254231ce1b76e912004475a38b9b20b798f8b6f6088c9762926322`
- The first is a zero-byte label-only child of the second; both have RootFS
  manifest SHA-256 `e52e6077145716853f7582208ceaef412ce4bbc1d05d4ce739a53b7f7a943f63`
- r17 base: `sha256:c5e96c5bcc5a073f7ce6b56173d88538de3a416900cff97c88b4bf7967fe1dc0`
- TP4 / DCP4 / MTP3 on four RTX PRO 6000 Blackwell GPUs
- dedicated `exl3_moe_r7_fused`; independent gate/up/down K3/K4/K5 tables
- FP32 sorted router probabilities
- `nvfp4_ds_mla` KV with calibrated static latent scales
- BF16 RoPE, BF16x3 router GEMM, `ONLINE_QUANT=none`
- `enforce_eager=False`; `FULL_AND_PIECEWISE` graphs captured on all ranks
- 463,872 physical KV tokens; 262,144-token model limit

### Performance

Measured with nominal 300 W caps per GPU, 1,200 W aggregate:

| Profile | Result |
|---|---:|
| Prefill 8K | 1,203 tok/s |
| Prefill 64K | 1,161 tok/s |
| Prefill 128K | 1,107 tok/s |
| Decode C1 | 74.0 tok/s |
| Decode C2 | 118.7 tok/s |
| Decode C4 | 192.5 tok/s |

### Five-run task battery

Measured post-reseat with nominal 300/300/300/275 W caps:

| Profile | Verdict | Mean output tokens | Aggregate generation |
|---|---:|---:|---:|
| LAVD | 5/5 correct (2 exact, 3 near) | 8,728.8 | 59.27 tok/s |
| Estonia | 5/5 pass | 1,744.2 | 55.65 tok/s |
| Hotel | 5/5 exact | 20,143.8 | 55.53 tok/s |
| Estonia-long | 5/5 pass | 1,951.4 | 54.82 tok/s |

No run hit its output-token limit. The very small TTFT values in this battery
are prefix-cache affected and are not presented as uncached prefill results.

### 260K retrieval

Five separate 260,000-token prompts retrieved the planted record exactly at
5%, 25%, 50%, 75%, and 95% depth. Elapsed times were 244.09, 239.77, 198.39,
136.18, and 70.02 seconds respectively. The decreasing times reflect growing
shared-prefix reuse and must not be read as an uncached-prefill comparison.

### Numerical evidence

Four comparable eager-harness BF16-reference KLD observations exist for the
exact configuration: `0.0639764262`, `0.0746190367`, `0.0704009716`, and
`0.0676657456`, with a descriptive mean of **0.0691655450**. They are one
sentinel plus three completed runs from a planned five-run panel. Run 4's log
ends during checkpoint loading without a recorded cause, and run 5 never
began; there is no five-run acceptance mean. The numerical KLD harness disabled
CUDA graphs. The separate serving/task/needle receipts establish graph replay.

The first NVFP4 observation, **0.0639764262**, is the lower NVFP4 result the
earlier discussion referred to (approximately `0.064`). It is a valid complete
2,047-position observation for the calibrated NVFP4/dedicated-R7/FP32-route
configuration, but it is not the multi-observation mean. Both the best observed
value and the four-observation descriptive mean are reported so neither is
lost or overstated.

### Hardware/error audit

- No request failures or output-limit hits
- No Xid, PCIe/AER, CUDA, traceback, or engine-fatal match
- No hardware-slowdown, thermal-slowdown, or power-brake samples
- Maximum observed temperature 89 C; C1 maximum 76 C
- C1's largest instantaneous telemetry sample was 372.29 W and largest
  averaged sample was 288.67 W, so the configured 275 W limit is stated as a
  nominal cap rather than a hard instantaneous clamp

### Source routing

- [vLLM PR #300](https://github.com/local-inference-lab/vllm/pull/300) provides
  the projection-mixed EXL3 loader and II-r17 serving foundation. The published
  narrow follow-up branch for cherry-pick onto #300 is
  [`brandonmmusic-max:vllm/codex/r17-b12x-without-r7-symbol`](https://github.com/brandonmmusic-max/vllm/tree/codex/r17-b12x-without-r7-symbol),
  with commits
  [`454b6b783045416e03c6a31731283523e4fc31e3`](https://github.com/brandonmmusic-max/vllm/commit/454b6b783045416e03c6a31731283523e4fc31e3),
  [`4aa039f66f3d25d52fc9f4f5b1c8e72b0c54141c`](https://github.com/brandonmmusic-max/vllm/commit/4aa039f66f3d25d52fc9f4f5b1c8e72b0c54141c), and
  [`239320cb9f84b24b66472d7a1acf77667d265f05`](https://github.com/brandonmmusic-max/vllm/commit/239320cb9f84b24b66472d7a1acf77667d265f05).
- Held local ExLlamaV3 kernel companion, based on the v0.0.43-era #246 head:
  `a5c57be199a74b970f74759501b6fa28c064e616`,
  `1c85f7602de42889129a593b442a6bfff55b6387`
- vLLM focused test file: 68 passed
- Independent ExLlamaV3 SM120a build: 68/68 units passed

The later source commits are intended post-validation reconstructions of the
effective overlays; exact equality to the builder source of the validated
images has not been independently proven. The original candidate labels record
local reviewed commit `b72c906...` and ExLlamaV3 base `704aefd...`, not the
later public commit IDs. The corresponding public vLLM commit has the identical
tree object. A new
self-contained image built successfully from the cumulative patch series and
passed the 68 focused tests, but has not yet repeated the end-to-end GPU
battery. No action is requested on ExLlamaV3 PR #246. That code is retained
locally for a possible later rebase and is outside this submission.

I have personally reviewed the proposed code changes, validation evidence, and
this submission, and I approve them for maintainer review. AI assistance was
used for implementation support, validation analysis, and drafting; I remain
responsible for the submitted code, results, and representations.
